api/v1alpha1/release_webhook.go

```diff
@@ -17,38 +17,116 @@ limitations under the License.
 package v1alpha1
 
 import (
-	"fmt"
-	"k8s.io/apimachinery/pkg/runtime"
+	"context"
+	"encoding/json"
 	"reflect"
+	"strings"
+
+	"gomodules.xyz/jsonpatch/v2"
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
-func (r *Release) SetupWebhookWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
-		Complete()
-}
+const authorLabel = "release.rhtap.openshift.io/author"
 
-//+kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-release,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=releases,verbs=create;update,versions=v1alpha1,name=vrelease.kb.io,admissionReviewVersions=v1
+// releaseHandler describes the handler for the Release low level webhook
+type releaseHandler struct {
+	Client client.Client
+}
 
-var _ webhook.Validator = &Release{}
+// +kubebuilder:webhook:path=/mutate-appstudio-redhat-com-v1alpha1-release,mutating=true,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=releases,verbs=create;update,versions=v1alpha1,name=mrelease.kb.io,admissionReviewVersions=v1
 
-// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
-func (r *Release) ValidateCreate() error {
-	return nil
+// SetupReleaseLowLevelWebhook registers the Release low level webhook with the passed manager.
+func SetupReleaseLowlevelWebhook(mgr ctrl.Manager) {
```

Comment:
Can you rename it to `Register` and pass `mgr ctrl.Manager` and `log *logr.Logger`? Save the logger in the struct you defined so it can be used in the `Handle` function.

This, btw, it's going to cause problems if you create a new webhook with the same function name. Ideally I'd have every crd (and webhooks) into a different sub-package, but I didn't managed to do it yet.

Reply:
done in follow up